L1Trigger/TrackFindingTracklet/src/Tracklet.cc

```diff
@@ -644,149 +647,204 @@ void Tracklet::setFitPars(double rinvfit,
   fpgatrack_ = std::make_unique<Track>(makeTrack(l1stubs));
 }
 
-std::string Tracklet::trackfitstr() {
-  string stubid0 = "111111111";
-  string stubid1 = "111111111";
-  string stubid2 = "111111111";
-  string stubid3 = "111111111";
+const std::string Tracklet::layerstubstr(const unsigned layer) const {
+  assert(layer <= 5);
+
+  std::stringstream oss("");
+  if (!layerresid_[layer].valid())
+    oss << "0|0000000|0000000000|0000000|000000000000|000000000";
+  else {
+    if (trackIndex_ < 0 || trackIndex_ > 127) {
+      cout << "trackIndex_ = " << trackIndex_ << endl;
+      assert(0);
+    }
+    const FPGAWord tmp(trackIndex_, 7, true, __LINE__, __FILE__);
+    oss << "1|";  // valid bit
+    oss << tmp.str() << "|";
+    oss << layerresid_[layer].fpgastubid().str() << "|";
+    oss << layerresid_[layer].stubptr()->r().str() << "|";
+    oss << layerresid_[layer].fpgaphiresid().str() << "|";
+    oss << layerresid_[layer].fpgazresid().str();
+  }
+
+  return oss.str();
+}
+
+const std::string Tracklet::diskstubstr(const unsigned disk) const {
+  assert(disk <= 4);
+
+  std::stringstream oss("");
+  if (!diskresid_[disk].valid())
+    oss << "0|0000000|0000000000|000000000000|000000000000|0000000";
+  else {
+    if (trackIndex_ < 0 || trackIndex_ > 127) {
+      cout << "trackIndex_ = " << trackIndex_ << endl;
+      assert(0);
+    }
+    const FPGAWord tmp(trackIndex_, 7, true, __LINE__, __FILE__);
+    const FPGAWord& stubr = diskresid_[disk].stubptr()->r();
+    oss << "1|";  // valid bit
+    oss << tmp.str() << "|";
+    oss << diskresid_[disk].fpgastubid().str() << "|";
+    oss << ((stubr.nbits() == 4) ? ("00000000" + stubr.str()) : stubr.str()) << "|";
+    oss << diskresid_[disk].fpgaphiresid().str() << "|";
+    oss << diskresid_[disk].fpgarresid().str();
+  }
+
+  return oss.str();
+}
+
+std::string Tracklet::trackfitstr() const {
+  string stub0;
+  string stub1;
+  string stub2;
+  string stub3;
+  string stub4;
+  string stub5;
+  string stub6;
+  string stub7;
+  string hitmap(24, '0');
```

Comment:
This function has lots of magic numbers. At the very least a comment is required explaining how they are derived. Ideally, they should be eliminated.

Reply:
Let me think about how best to clean this method up.

Reply:
@tomalin I updated this method to be more readable and include more comments. And I eliminated as much magic as I could without introducing revolutionary changes.